atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/ConflictDetectionManager.java

```diff
@@ -15,56 +15,133 @@
+    private static final Logger log = LoggerFactory.getLogger(ConflictDetectionManager.class);
+    private final LoadingCache<TableReference, ConflictHandler> cache;
 
     /**
-     * @param supplier A {@link Supplier} which returns the current conflict detection
-     *                 configuration to the database. Note that the result is cached,
-     *                 so changes to configuration must be followed by a call to
-     *                 {@link #recompute()} in order to be picked up.
+     *  This class does not make the mistake of attempting cache invalidation,
+     *  so a table dropped by another instance may still be cached here.
+     *
+     *  This is okay in the case of a simple drop, but a same-name table drop
+     *  and re-addition with a different Conflict Handler
+     *  (where an external atlas instance handles both of these operations)
+     *  will be incorrect. This is an unrealistic workflow
+     *  and I'm fine with just documenting this behavior.
+     *
+     *  (This has always been the behavior of this class; I'm simply calling it out)
      */
-    ConflictDetectionManager(Supplier<Map<TableReference, ConflictHandler>> supplier) {
-        this.supplier = supplier;
-        this.overrides = Maps.newConcurrentMap();
-        this.cachedResult = RecomputingSupplier.create(() -> {
-            Map<TableReference, ConflictHandler> ret = new HashMap<>(this.supplier.get());
-            ret.putAll(overrides);
-            return Collections.unmodifiableMap(ret);
-        });
+    private ConflictDetectionManager(CacheLoader<TableReference, ConflictHandler> loader) {
+        this.cache = CacheBuilder.newBuilder()
+                .maximumSize(100_000)
+                .build(loader);
     }
 
-    public void setConflictDetectionMode(TableReference table, ConflictHandler handler) {
-        overrides.put(table, handler);
-        cachedResult.recompute();
+
+    public static ConflictDetectionManager create(KeyValueService kvs) {
+        ConflictDetectionManager conflictDetectionManager = new ConflictDetectionManager(
+                new CacheLoader<TableReference, ConflictHandler>() {
+                    @Override
+                    public ConflictHandler load(TableReference tableReference) throws Exception {
+                        byte[] metadata = kvs.getMetadataForTable(tableReference);
+                        if (metadata == null) {
+                            return ConflictHandler.RETRY_ON_WRITE_WRITE; // legacy behavior / this is dumb
+                        }
+                        return getConflictHandlerFromMetadata(metadata);
+                    }
+                });
+
+        // kick off an async thread that attempts to fully warm this cache
+        // if it fails (e.g. probably this user has way too many tables), that's okay,
+        // we will be falling back on individually loading in tables as needed.
+        new Thread(() -> {
```

Comment:
This seems like an overkill. Loading metadata for a table should be pretty cheap.